### ISSUE TITLE
Assist computation is way too slow

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#15368]
Vague issue, but I noticed this somewhat frequently in some files that assists take a really long time to compute and show up at times. So far we haven't really given much thought to this aside from trying to make them somewhat lazy but it seems there are some assists that do a lot of upfront work still, we should investigate that.

**Comment by matklad** [comment]
Yeah, I think assist interface, with lazy/eager part is in general right. What’s hard is using this interface correctly, assist authors don’t necessary realize that they might be triggering costly searches from the upfront work.

I wonder if we need to flesh out our “slow event loop” infra? I’ve noticed that Deno extension has a nice “status” page as a markdown document (server emits markdown, vscode then renders tables and such). What we could do here is to have an extended Status page for ra, which would record slowest assists, highlights, etc. that is, you, as a developer of ra, open the status page, and see “aha, ten minutes ago auto-import in foo.rs at offsets 42 took 5 seconds, gotta look into that”.
